### PR TITLE
feat: [CDS-81118]: runtime support for single remote service and remote multi services 

### PR DESCRIPTION
**Pull request by vikrant-harness**
### Summary

- Runtime Support for remote single service and remote multi services in run pipeline form / Input set and templates 
- setting up parent pipeline context in run pipeline form and storing it in stage form context

#### Screenshots

https://github.com/harness/harness-core-ui/assets/63278928/f9af587c-7bee-4397-b08f-9cfa4ddff0f6

https://github.com/harness/harness-core-ui/assets/106532291/847579b9-29e0-4bf8-b6ed-28a139004c12

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
- MessageMetadata: `trigger messagecheck`
- GitLeaks: `trigger gitLeaks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

### CODE REVIEW

File: src/modules/70-pipeline/context/StageFormContext.tsx [hunk 1]
```diff
@@ -12,6 +12,11 @@ import type { PipelineInfoConfig } from 'services/pipeline-ng'
 export interface StageFormContextValues {
   getStageFormTemplate<T>(path: string): T | PipelineInfoConfig
   updateStageFormTemplate<T = unknown>(data: T, path: string): void
+  pipelineGitMetaData?: {
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/context/StageFormContext.tsx:15]:
```
type PipelineGitMetaData = {
  branch: string
  repoName: string
  connectorRef: string
}
```
```suggestion
  pipelineGitMetaData?: PipelineGitMetaData
```
And the use PipelineGitMetaData at all other places where these values need to be passed

File: src/modules/70-pipeline/components/PipelineInputSetForm/ChainedPipelineInputSetUtils.tsx [hunk 2]
```diff
@@ -36,6 +36,9 @@ export interface ChainedPipelineInputSetFormProps {
   maybeContainerClass?: string
   viewTypeMetadata?: Record<string, boolean>
   disableRuntimeInputConfigureOptions?: boolean
+  repoIdentifier?: string
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineInputSetForm/ChainedPipelineInputSetUtils.tsx:39]:
```
export interface ChainedPipelineInputSetFormProps extends Optional<PipelineGitMetaData> {
}
```

File: src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx [hunk 3]
```diff
@@ -80,6 +80,9 @@ export interface PipelineInputSetFormProps {
   stageTooltip?: {
     [key in StageType]?: string
   }
+  branch?: string
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx:83]: Same as above

File: src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx [hunk 4]
```diff
@@ -217,6 +223,9 @@ export function StageForm({
   stageTooltip?: {
     [key in StageType]?: string
   }
+  repoIdentifier?: string
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx:226]: Same as above

File: src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironmentEntityInputStep.tsx [hunk 5]
```diff
@@ -74,6 +75,11 @@ export default function DeployEnvironmentEntityInputStep({
   const { getStageFormTemplate, updateStageFormTemplate } = useStageFormContext()
   const formik = useFormikContext<DeployEnvironmentEntityConfig>()
   const uniquePath = useRef(`_pseudo_field_${uuid()}`)
+  const {
+    state: { storeMetadata }
+  } = usePipelineContext()
+
+  const { pipelineGitMetaData } = useStageFormContext()
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironmentEntityInputStep.tsx:82]: Both of them must have same data. Why do we have to rely on both?

File: src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityInputStep.tsx [hunk 6]
```diff
@@ -61,7 +63,10 @@ export function DeployServiceEntityInputStep({
   const { getString } = useStrings()
   const { showWarning } = useToaster()
   const { expressions } = useVariablesExpression()
-  const { updateStageFormTemplate } = useStageFormContext()
+  const { updateStageFormTemplate, pipelineGitMetaData } = useStageFormContext()
+  const {
+    state: { storeMetadata }
+  } = usePipelineContext()
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityInputStep.tsx:69]: Same as above.

File: src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityUtils.ts [hunk 7]
```diff
@@ -29,6 +29,7 @@ export interface DeployServiceEntityData {
     metadata?: ServicesMetadata
     values?: ServiceYamlV2[] | string
   }
+  serviceGitBranches?: { [x: string]: string }
```
Comment (chetanvekariya-harness) [inline src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityUtils.ts:32]: nit: { [key: string]: string }

**Review by chetanvekariya-harness — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityUtils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (chetanvekariya-harness) [reply src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityUtils.ts:32]: +1

File: src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironmentEntityInputStep.tsx (hunk 5 above)
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/DeployEnvironmentEntityStep/DeployEnvironmentEntityInputStep.tsx:82]: The pipeline studio context is not present in the run pipeline form so we need stage context for the same

File: src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityInputStep.tsx (hunk 6 above)
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityInputStep.tsx:69]: The pipeline studio context is not present in the run pipeline form so we need stage context for the same

**Review by sunnykesh-harness — APPROVED** [review]

File: src/modules/70-pipeline/context/StageFormContext.tsx (hunk 1 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/context/StageFormContext.tsx:15]: Done

File: src/modules/70-pipeline/components/PipelineInputSetForm/ChainedPipelineInputSetUtils.tsx (hunk 2 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineInputSetForm/ChainedPipelineInputSetUtils.tsx:39]: done

File: src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx (hunk 3 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx:83]: Done

File: src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx (hunk 4 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx:226]: Done

File: src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityUtils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityUtils.ts:32]: done

**Review by chetanvekariya-harness — APPROVED** [review]

File: src/modules/70-pipeline/components/RunPipelineModal/VisualView.tsx [hunk 8]
```diff
@@ -261,6 +266,9 @@ export interface PipelineInputSetFormWrapperProps {
   showDivider?: boolean
   listOfSelectedStages?: string[]
   isRetryFormStageSelected?: boolean
+  repoIdentifier?: string
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/RunPipelineModal/VisualView.tsx:269]:
```suggestion
  repoName?: string
```

then use the same Interface here as well

**Review by sunnykesh-harness — APPROVED** [review]

File: src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx [hunk 9]
```diff
@@ -217,6 +220,9 @@ export function StageForm({
...
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx:223]:
```suggestion
  repoName?: string
```

then use same interface here as well

File: src/modules/70-pipeline/components/RunPipelineModal/VisualView.tsx (hunk 8 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/RunPipelineModal/VisualView.tsx:269]: Will cleanup in follow up PR

File: src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx (hunk 9 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineInputSetForm/PipelineInputSetForm.tsx:223]: Will cleanup in follow up PR

**Pull request merged**